Feature/kas 4199 editable pieces with sign flows

sergiofenoll opened the pull request:
Show the dropdown for pieces that have a MARKED sign flow. Delete and re-creates sign flows as needed.

sergiofenoll committed:
feat: show piece edit dropdown when a marked sign flow exists

We still hide it for any sign flows that's further along (i.e. on SH).
We also ensure that the dropdown only renders after we've loaded the
signature data. This to prevent some content flashing, as before the
dropdown would show for a split second while the sign flow data was
being loaded.

sergiofenoll committed:
feat: delete and re-create sign flows as needed when changing pieces

sergiofenoll committed:
feat: extract updatePreviousAccessLevel from document-card

At least in the agendaitem view, we are refreshing the page after adding
a piece. Although up until now the call to update the access levels
that's in the document card seems to have gone through (even though the
component where the call originates from was re-rendered!) it seems
safer to extract this call. We're leaving in the document-card just in
case though.

sergiofenoll committed:
feat: only show dropdown for marked pieces with a decision activity

In practice this means that the piece will only be editable in the
agendaitem documents route. This is because from e.g. the subcase view
we don't directly have the decision activity, which means that something
like uploading a new version (which means we need to create a new sign
flow) will fail. So we just prevent the behaviour instead.

sergiofenoll commented on app/components/documents/document-card.js:
```diff
@@ -74,6 +76,17 @@ export default class DocumentsDocumentCardComponent extends Component {
+      && this.loadSignatureRelatedData.isIdle
+      && this.loadSignatureRelatedData.performCount > 0
+      && (!this.hasSignFlow
+          || (this.hasMarkedSignFlow && !!this.args.decisionActivity))
```
This condition is to stop us from showing the dropdown in the `subcase/documents` route. Since we don't have a decisionActivity there, we can't do `Nieuwe versie opladen`, because we can't re-create the sign flow.

We might instead want to only show the options that are possible to execute (`Wijzigen` and `Document verwijderen`).

tdn replied:
Yes, that's probably a good idea, since the users were not happy about "losing" their features.

sergiofenoll replied:
:+1: Changes made, though thinking about it I guess we could still show `Nieuwe versie opladen` in the subcase route by getting the decisionActivity from the old sign flow (before deleting it) and then re-using it to create the new one.

I can make that change too if desired, otherwise we can keep it in the back of our minds if the question ever comes from the users.

ValenberghsSven replied:
I would wait for an explicit ask from the users.
I've heard numerous times that "the agenda view" is where secretary will be marking the documents.
If the marking has been toggled then that would have happened in the agenda views.
It makes more sense that adding new versions of documents with markings would also be done in the agenda views.
No dropdown/actions on subcase views docs when a markingActivity is present for now seems ok.

edit:
Edit document and delete document container allowed should be ok when it is only marked.

sergiofenoll committed:
feat: check that we still have marked sign flow before mutating the piece

ValenberghsSven commented on translations/nl-be.json:
```diff
@@ -1248,5 +1248,6 @@
   "to-follow-up": "Op te volgen",
-  "stop-sign-flow": "Tekenen stopzetten"
+  "stop-sign-flow": "Tekenen stopzetten",
+  "sign-flow-was-sent-while-you-were-editing": "De sign flow is reeds verstuurd naar de Handtekenmap terwijl u probeerde te editeren."
```
This translation is most likely too technical to show to the user.
"Dit document is reeds verstuurd ... "?
We also don't specifiy that because of this we block the upload of a new version of a document

sergiofenoll replied:
The toast title should be `Je wijzigingen kunnen niet opgeslagen worden.`, not sure if that's enough for users, but maybe explicitly telling them that uploads/edits/deletes weren't saved is better. Will change and make it simpler.

ValenberghsSven reviewed (approved):
Code looks fine, have not tested it manually but jenkins is green so that helps.

Minor comment on the translation but is not blocking for approval

sergiofenoll committed:
chore: add better error messages and don't await refresh

With the awaited refresh, the toasts wouldn't render.

sergiofenoll committed:
feat: be more lenient with edit dropdown, but hide actions contextually

Uploading a new version of a piece with a sign flow in the subcase route
is now hidden (because we lack the decision activity to recreate the
sign flow), but the rest of the dropdown is still available.

ValenberghsSven committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-4199-editable-pieces-with-sign-flows

ValenberghsSven commented on app/components/documents/document-card.js:
```diff
@@ -202,8 +226,21 @@ export default class DocumentsDocumentCardComponent extends Component {
+      const status = yield this.signFlow.belongsTo('status').reload();
+      if (status.uri !== CONSTANTS.SIGNFLOW_STATUSES.MARKED) {
+        yield this.deleteUploadedPiece.perform();
+        this.router.refresh();
```
A refresh from component? Should be the responsibility of controllers/routes.
Also, why are we refreshing here?
We could just do `deleteUploadedPiece` and show the warning.

sergiofenoll replied:
Refresh aren't strictly needed, but it's to force the user to see the page with the new data. To prevent them mindlessly dismissing the toast and just keep trying the action and such. As long as the page isn't refreshed we're still in a stale/inconsistent state.

ValenberghsSven replied:
tested this locally and the only issue is that the actions dropdown stays visible after we reloaded the status.
So running `loadSignatureRelatedData` again is enough (only 1 card needs to be refreshed, not the entire route)

ValenberghsSven commented on app/components/documents/document-card.js:
```diff
@@ -241,7 +278,19 @@ export default class DocumentsDocumentCardComponent extends Component {
+    if (this.signFlow) {
+      const status = await this.signFlow.belongsTo('status').reload();
+      if (status.uri !== CONSTANTS.SIGNFLOW_STATUSES.MARKED) {
+        this.router.refresh();
```
same here

ValenberghsSven commented on app/components/documents/document-card/edit-modal.js:
```diff
@@ -92,6 +94,19 @@ export default class DocumentsDocumentCardEditModalComponent extends Component {
+    if (this.args.signFlow) {
+      const status = await this.args.signFlow.belongsTo('status').reload();
+      if (status.uri !== CONSTANTS.SIGNFLOW_STATUSES.MARKED) {
+        this.router.refresh();
```
just call the `onCancel` and call it a day? no router refresh needed in component

ValenberghsSven replied:
called local `this.cancelEdit()` which also deletes any uploaded files and call `args.onCancel`

ValenberghsSven reviewed (commented):
Missed it on my first run through. Refresh from component should not be needed.
Should be the concern of the controller with callbacks if that is needed.

tdn commented on app/services/signature-service.js:
```diff
@@ -206,6 +206,18 @@ export default class SignatureService extends Service {
 
+  async hasMarkedSignFlow(piece) {
+    const signaturesEnabled = !!ENV.APP.ENABLE_SIGNATURES;
+    if (signaturesEnabled) {
```
I realize this was already the case in `hasSignFlow`, but just a consideration I'm making now: if we turn the feature flag off after we've created some `signFlows`. That will mean all edit functionalities become available again, and we could end up with `signFlows` that have different/no pieces in de database.

Not something to fix in this PR, but something to remember if we ever enable and then disable the feature flag: we need a "cleanup" function to check for inconsistent `signFlows`.

ValenberghsSven replied:
if we are willingly turning off the feature flag for any reason we should just do a full cleanup of the signing graph + data?
In time the feature flag should be removed and the feature fully integrated in the app.

ValenberghsSven replied:
If we have this feature running on prod once we should never disable it in my opinion.
They can just choose not to use it at that stage (but automatic things will keep happening on marked pieces)

tdn reviewed (changes requested):
This fixes the issue for the document card, but not for the "Bulk edit" modal or for the DocumentPreviewSidebar.

As I understand it, @ValenberghsSven is taking care of the "Bulk edit" functionalities already due to conflicts with his PR.
He told me he'll pick both of these up.

ValenberghsSven committed:
KAS-4199 handle sad flows in component with reload instead of router refresh

ValenberghsSven commented on app/components/documents/document-card.js:
```diff
@@ -311,4 +365,11 @@ export default class DocumentsDocumentCardComponent extends Component {
+
+  @action
+  async cancelEditPiece() {
+    await this.loadPieceRelatedData.perform();
```
this `loadPieceRelatedData` already contains the `loadSignatureRelatedData`.
Just noticed that now.
It might be better to just load that one. Will adjust the other ones I just added

ValenberghsSven committed:
KAS-4199 use full reload on sad flow

ValenberghsSven committed:
KAS-4199 add editing and deleting to marked sign flows

ValenberghsSven committed:
KAS-4199 add used services

ValenberghsSven commented:
did not fully test the edit/deletion of document route with a marked piece / concurrent status change.

ValenberghsSven committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-4199-editable-pieces-with-sign-flows

tdn commented:
Closing this since this branch was merged into https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1831

Will be merged into development from there
